### DISCUSSION TITLE
How Can I launch an experiment using Golang/Python?

### DISCUSSION THREAD

**Discussion by cauwulixuan** [iter8-tools/iter8#1265] [Q&A] [answered]
I'm newbee here, I followed [Tutorials](https://iter8.tools/0.10/getting-started/your-first-experiment/#your-first-experiment) here, and here is my question:

In v0.7, I can setup CRDs in kubernetes cluster, and launch experiments by setup Experiment CR, how can I do that in v0.10?

If I need launch experiments and report results in my app (written by golang or python ...), how can I make it work using a iter8 SDK(something like that) OR just setup CRD and CR instead of a binary `iter8` CLI ?

Many thanks if you have any suggestions.

**Comment by sriumcp** [comment, marked as answer]
Hi there! Welcome to Iter8!
***
In v0.10, you do not need a CRD to run experiments inside Kubernetes. Iter8 v0.10 uses a combination of namespace scoped built-in Kubernetes resources (like secrets and jobs) to run the experiment inside Kubernetes.
***
The `iter8` binary enables you to set up Kubernetes experiments in addition to local experiments. Please refer to https://iter8.tools/0.10/tutorials/load-test-http/kubernetesusage/ and https://iter8.tools/0.10/tutorials/load-test-grpc/kubernetesusage/. Both of these are examples of launching experiments inside Kubernetes.
***
In the Golang source for Iter8, launching Kubernetes experiments is enabled by [this method](https://github.com/iter8-tools/iter8/blob/53f174df0ade094171808028b559aaa74c107c7e/action/launch.go#L93). For an example of how this method is used, please see [here](https://github.com/iter8-tools/iter8/blob/53f174df0ade094171808028b559aaa74c107c7e/cmd/klaunch.go#L41).

Perhaps the Golang SDK for launching experiments should be better documented in the README? If you think this documentation would help, could you please create a documentation issue in this repo? We are happy to work on these docs if it would be useful for you.

**Reply by cauwulixuan** [reply]
Thanks for your reply @sriumcp 

> Iter8 v0.10 uses a combination of namespace scoped built-in Kubernetes resources (like secrets and jobs) to run the experiment inside Kubernetes.

I can see `secrets` and `jobs` were created after I ran
```
iter8 k launch -c load-test-http --set url=https://httpbin.org/get
```
But I can't just use `iter8 CLI` in my code, right? Do you have any other more native usage in my senario? I also create a documentation issue #1267

**Comment by cauwulixuan** [comment]
@sriumcp BTW, How can I setup custom metrics? I want to use several custom metrics in prometheus, not only built-in HTTP latency and error-related metrics.  Should I setup a new task, and could you provide a official progress on setup `task`?

**Reply by sriumcp** [reply]
Hi @cauwulixuan Yes, custom Prometheus metrics (or any other database) is already possible in v0.10. We're in the process of documenting this. @Alan-Cha... comments?
